RecoEgamma/ElectronIdentification/plugins/ElectronMVANtuplizer.cc

```diff
@@ -0,0 +1,534 @@
+          mvaValues_[k] = (*values[k])[ele];
+        }
+
+        for (size_t k = 0; k < nCats_; ++k) {
+          mvaCats_[k] = (*mvaCats[k])[ele];
+        }
+
+
+        tree_->Fill();
+    }
+
+}
+
+void ElectronMVANtuplizer::findFirstNonElectronMother2(const reco::Candidate *particle,
+                         int &ancestorPID, int &ancestorStatus){
+
+  if( particle == nullptr ){
+    printf("ElectronNtupler: ERROR! null candidate pointer, this should never happen\n");
+    return;
+  }
+
+  // Is this the first non-electron parent? If yes, return, otherwise
+  // go deeper into recursion
+  if( abs(particle->pdgId()) == 11 ){
+    findFirstNonElectronMother2(particle->mother(0), ancestorPID, ancestorStatus);
+  }else{
+    ancestorPID = particle->pdgId();
+    ancestorStatus = particle->status();
+  }
+
+  return;
+}
+
+template<class T, class V>
+int ElectronMVANtuplizer::matchToTruth(const T &el, const V &prunedGenParticles, int &genIdx){
+
+  //
+  // Explicit loop and geometric matching method (advised by Josh Bendavid)
+  //
+
+  // Find the closest status 1 gen electron to the reco electron
+  double dR = 999;
+  const reco::Candidate *closestElectron = nullptr;
+  for(size_t i=0; i<prunedGenParticles->size();i++){
+    const reco::Candidate *particle = &(*prunedGenParticles)[i];
+    // Drop everything that is not electron or not status 1
+    if( abs(particle->pdgId()) != 11 || particle->status() != 1 )
+      continue;
+    //
+    double dRtmp = ROOT::Math::VectorUtil::DeltaR( el->p4(), particle->p4() );
+    if( dRtmp < dR ){
+      dR = dRtmp;
+      closestElectron = particle;
+      genIdx = i;
+    }
+  }
+  // See if the closest electron (if it exists) is close enough.
+  // If not, no match found.
+  if( !(closestElectron != nullptr && dR < deltaR_) ) {
+    return UNMATCHED;
+  }
+
+  //
+  int ancestorPID = -999;
+  int ancestorStatus = -999;
+  findFirstNonElectronMother2(closestElectron, ancestorPID, ancestorStatus);
+
+  if( ancestorPID == -999 && ancestorStatus == -999 ){
+    // No non-electron parent??? This should never happen.
+    // Complain.
+    printf("ElectronNtupler: ERROR! Electron does not apper to have a non-electron parent\n");
+    return UNMATCHED;
+  }
+
+  if( abs(ancestorPID) > 50 && ancestorStatus == 2 )
+    return TRUE_NON_PROMPT_ELECTRON;
+
+  if( abs(ancestorPID) == 15 && ancestorStatus == 2 )
+    return TRUE_ELECTRON_FROM_TAU;
+
+  // What remains is true prompt electrons
+  return TRUE_PROMPT_ELECTRON;
+}
+
+// ------------ method called once each job just before starting event loop  ------------
+void
+ElectronMVANtuplizer::beginJob()
+{
+}
+
+// ------------ method called once each job just after ending the event loop  ------------
+void
+ElectronMVANtuplizer::endJob()
+{
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void
+ElectronMVANtuplizer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
```

Comment:
Add descriptions here and remove the ```existsAs``` above

Reply:
Should I only add the parameters which had a default and were checked with `existsAs` or all of them?

Reply:
Please use the fillDescriptions method to define all parameters
The same can be done also later in the PhotonMVANtuplizer: I see no reason why to leave the method dummy

Reply:
Ok, done. The other comments which are now marked as outdated were all addressed too.